Clean/Circuit/Theorems.lean

```diff
@@ -189,33 +189,46 @@ theorem can_replace_local_witnesses_completeness {env: Environment F} {n: ℕ} {
     apply circuit.implied_by_local_witnesses
     rw [←extends_vector_subcircuit]
     exact h.right
+
+theorem uses_local_witnesses_completeness_iff_forAll {env: Environment F} {n: ℕ} {ops: Operations F n} :
+  env.uses_local_witnesses_completeness ops ↔
+    ops.forAll {
+      witness n _ c := env.extends_vector (c env) n,
+      assert _ _ := True,
+      lookup _ _ := True,
+      subcircuit _ s := s.uses_local_witnesses env
+    } := by
+  induction ops with
+  | empty => trivial
+  | assert | lookup | witness | subcircuit =>
+    simp_all [uses_local_witnesses_completeness, Operations.forAll]
 end Environment
 
 namespace Circuit
-/--
-Generic version of `constraints_hold`, to reason about soundness and completeness at the same time
--/
-def constraints_hold.generic (from_subcircuit : {n : ℕ} → Environment F → SubCircuit F n → Prop)
-  (eval : Environment F) {n : ℕ} : Operations F n → Prop
-  | .empty _ => True
-  | .witness ops _ _ => generic from_subcircuit eval ops
-  | .assert ops e => generic from_subcircuit eval ops ∧ eval e = 0
-  | .lookup ops { table, entry, .. } => generic from_subcircuit eval ops ∧ table.contains (entry.map eval)
-  | .subcircuit ops s => generic from_subcircuit eval ops ∧ from_subcircuit eval s
```

Comment:
this wasn't generic enough, as already noted by @gio54321 in the initial review: https://github.com/Verified-zkEVM/clean/pull/71#discussion_r2042210357